### PR TITLE
Move mode-enter manual throttle checks up

### PR DESCRIPTION
**Pull request by peterbarker**

### CODE REVIEW

File: ArduCopter/mode.cpp [hunk 1]
```diff
@@ -200,6 +200,20 @@ bool Copter::set_mode(control_mode_t mode, mode_reason_t reason)
     }
 #endif
 
+    // ensure vehicle doesn't leap off the ground if a user switches
```
Comment (rmackay9) [inline ArduCopter/mode.cpp]: I think we should add #if FRAME_CONFIG != HELI_FRAME around this check

Comment (peterbarker) [reply ArduCopter/mode.cpp]: Added.

**Comment by ChristopherOlson** [comment]
@bnsgeyer brought this to my attention so I reviewed it.

We have a problem here.

Rule of thumb for helicopters; the pilot must ALWAYS be able to assume instant and full manual control of the machine at any time, no matter what the situation is.

The entire system does not deal well with helicopters on the ground in the first place. So we're creating a Boeing 737 Max problem in ArduPilot with a black box thinking it knows more than the pilot.

And further, the collective is not the same as the throttle on multicopters. That is just an unfortunate design in the code. Some heli setups have zero pitch at 50% collective travel. Which means that helicopter can be run up to full power on the ground at 50% travel and the pilot has not started to pull pitch yet in a manual mode. If the pilot starts to pull pitch and the machine goes into ground resonance, he/she tries to switch to a manual mode to get it out of there before it explodes, this would prevent her from taking manual control of the machine from the autopilot. And that's simply unacceptable. They don't respond to the controls properly in autopilot modes. If the autopilot loses it on the ground and it has a blade strike and blows a pitch link, the machine turns into a deadly missile. The only way out to be able to switch to the manual mode.

**Comment by rmackay9** [comment]
@ChristopherOlson, Peter's adding the definitions in so that heli isn't affected.

I wouldn't jump to the conclusion though that TradHeli user's wouldn't benefit from the check working as it does on multicopters.  ArduPilot users come in all different levels of experience and this safety check has very likely saved some multicopter users from unintended takeoffs.

**Comment by peterbarker** [comment]
@ChristopherOlson As requested on the devcall this-morning, the check in question is guarded against being  crossed by heli: https://github.com/ArduPilot/ardupilot/pull/10720/files#diff-bc5b258afd514aa95767c65b4b95e90bR203  .  Given that, where is the problem for Heli?

In testing we don't actually cross it anyway, as the rotor runup must be complete and `ap.land_complete` goes false.

Further testing shows that If you're flying a heli in `GUIDED` mode, on the ground with your collective at maximum and rotor runup complete, switching to `STABILIZE` launches the vehicle into the air (on a real vehicle, possibly rather rapidly).  I imagine that's "working as intended", however, given your comments above.

**Comment by ChristopherOlson** [comment]
Peter, that's correct. That's what it should do if the pilot switches it. The reason it's going to get switched is because the autopilot screws up and the pilot then realizes this was a Very Bad Idea. At that point you're either going wad the helicopter up or get it out of there before the autopilot tips it over.

Like Randy said, users come in all experience levels. Do the above once, and you'll learn that's not the thing to do. That's why most users will start out with a little one to learn the ropes in the hope of being able to gain enough experience to fly a big one. Nobody I know of has every bought a 700+ class helicopter and flown it successfully without either training or lots of experience, regardless of whether or not it is equipped with an autopilot.

So we have different clientele in the big picture and I'm always going to have a different outlook because I fly the big ones you ride in. There's things an autopilot should do, and things it should not be doing. This one is the latter. Heli doesn't need to be included in it at all. That's why it was never in Copter::ModeAcro_Heli::init(bool ignore_checks) in acro, but it might seem like the best thing ever in multirotor acro init()

**Comment by ChristopherOlson** [comment]
@peterbarker not related to this PR but there's also some other annoyances for heli due to things that have gotten into Copter for multi's and one of those is the arming check for collective position. Can't arm a heli controller with the collective off the bottom stop. Pre-flight checks in a heli is to always zero the pitch after cyclic checks before starting the engine. Zero pitch is never at the bottom stop in RC helicopters.

Got the second complaint on that yesterday from a user. That is another thing that is on my List of Things to Fix  :-)

**Comment by peterbarker** [comment]
@ChristopherOlson So you're saying our prearms currently state we have to have full down collective when it should be neutral?  On the face of it that does sound wrong to me...

**Comment by bnsgeyer** [comment]
@peterbarker I can see Chris’ point on this. When it comes to arming, there are checks that don’t apply to heli. This is one of them. It is due to the fact that  The motors are not engaged on arming if the collective is not on the bottom stop. Unlike multi copter where if the throttle was not on the bottom stop your aircraft could spool up and take off.  For helicopters we have the motor interlock pre-arm check that is our safety feature.  The collective pre-arm check could be done away with for Heli’s or any aircraft using the motor interlock switch.

**Comment by ChristopherOlson** [comment]
Yeah, it does this. We have to fix it. And not check for pitch neutral. It needs to be removed from the code.

When changes are deemed necessary whatever the behavior is for heli should not be altered after the change is made just because it looks like a "good idea". As in this case, especially in acro, it was not an oversight - it's by design. Sim testing tells little about behavior with heli's unless you test a specific setup, and the different setups for heli's are virtually infinite. The manual collective control modes are PRIMARY for helicopter, not secondary.

The mod looks ok to me since it doesn't apply to heli frame. I can see why it's desirable to move it up higher in the code instead of having it in the flight modes. But in this case heli had it's own init() and run() for acro so I'm glad it got caught before it became a problem. The timing is not right for us to always make the Dev Call, as that's the time of day I'm normally pushing helicopters or airplanes into the hanger and I'm still at the airport. It's morning in Australia, it's the end of the work day here.

**Pull request merged**